src/controllers/discordController.ts

```diff
@@ -0,0 +1,176 @@
+            res.status(400).send('A user with that email already exists. Please log in with the method used to sign up, and link your Discord account in the settings.');
+        } else if(existingdiscordUser) {
+            res.status(400).send('This Discord account is already linked to an account. Please log in with that account.');
+        } else {
+            const newUser = userRepo.create({
+                uuid: uuidv4(),
+                role: Role.Hacker,
+                email: response.data.email,
+                discordId: response.data.id
+            } as User);
+
+            const errors = await validate(newUser);
+            if(errors.length > 0) {
+                res.sendStatus(400);
+            } else {
+                try {
+                    await userRepo.save(newUser);
+                    // Login the new user
+                    const accessToken = await auth.generateToken(newUser.uuid, 'access');
+                    const refreshToken = await auth.generateToken(newUser.uuid, 'refresh');
+                    res.status(200).send({accessToken: accessToken, refreshToken: refreshToken});
+                } catch (error) {
+                    res.status(400).send(error);
+                }
+            }
+        }
+    }
+}
+
+export const discordLinkCallback = async (req: Request, res: Response): Promise<void> => {
+    // Before we do anything, check that the state is valid
+    const state = req.query.state as string | undefined;
+    const stateRepo = getManager().getRepository(State);
+    const savedState = await stateRepo.findOne({state: state, type: 'discord'});
+
+    if(!savedState) {
+        res.sendStatus(400);
+        return;
+    }
+    await stateRepo.remove(savedState); // We don't need this state anymore
+    
+    // For some reason Discord needs a redirectUri here
+    const discordUser = await discordAuth.code.getToken(req.originalUrl, {redirectUri: 'http://localhost:8080/api/auth/discord/callback/link'});
+    const url = discordUser.sign({method: 'get', url: 'https://discord.com/api/users/@me'}) as AxiosRequestConfig;
+    const response = await axios(url);
+
+    const userRepo = getManager().getRepository(User);
+    const userUuid = state?.split('/')[1];
+
+    if(!userUuid) {
+        res.status(400).send('Invalid state.'); // The state is wrong...
+    } else {
+        const existingdiscordUser = await userRepo.findOne({discordId: response.data.id});
+        if(existingdiscordUser){
+            res.status(400).send('This Discord account is already linked to an account. Please log in with that account.');
+            return;
+        }
+
+        const user = await userRepo.findOne({uuid: userUuid});
+        if(!user) {
+            res.sendStatus(400); // Can this even happen???
+        } else {
+            if(user.discordId) {
+                res.status(400).send('This user already has a Discord account linked. Please unlink first, then try again.'); // Do we want this?
```

Comment:
What would be the user flow that leads to this?

Reply:
From what I can tell, this can only happen if somehow, you got a valid state with an invalid uuid. Which should be impossible, since only we can generate states. This if statement should also catch any issues with TypeORM getting that user for any reason though, so it's still a good check to have.